### ISSUE TITLE
[InstCombine] missing canonicalizations for bitwise logic + add

### ISSUE THREAD

**Issue by rotateright** [llvm/llvm-project#56731]
We changed a canonicalization for "and + add" in:
https://reviews.llvm.org/D130080
...which suggests that we should have canonicalizations for the other bitwise logic opcodes too for consistency:
or + add:
https://alive2.llvm.org/ce/z/SQNt-L
xor + add:
https://alive2.llvm.org/ce/z/64RNkZ

**Comment by emgullufsen** [comment]
I am hoping to submit a patch for these other math + logic combinations. One initial question - a comment comment in [D130080](https://reviews.llvm.org/D130080) says :
```
// If all bits affected by the add are included in a high-bit-mask, do the
// mask op before the add.
```
So it sounds like the matching logic will be pretty similar to handle these other combinations, but placed in the methods that handle or and xor combinations?

This is my second patch so I hope to need much much less hand-holding this time - thank you!

**Comment by emgullufsen** [comment]
Sorry - I believe I misunderstood. From the CompilerExplorer examples in the issue description, it looks as if we want to flip the sequence "xor followed by add" to "add followed by xor" (likewise for "or, add" -> "add, or") provided that the logic operation does not affect the same bits as the math op (add).

**Comment by rotateright** [comment]
> Sorry - I believe I misunderstood. From the CompilerExplorer examples in the issue description, it looks as if we want to flip the sequence "xor followed by add" to "add followed by xor" (likewise for "or, add" -> "add, or") provided that the logic operation does not affect the same bits as the math op (add).

Right - D130080 is proposing to move the bitwise logic first, so the alive2 proofs in this report are inverted. You can add "--bidirectional" to alive2 code to show that a transform is reversible:  
https://alive2.llvm.org/ce/z/T8DvcR

I see that hasn't been pushed yet, so you probably want to wait for that to happen.

For implementation, you would probably create a small helper function that takes the logic opcode as a parameter, and then call that helper from visitAnd / visitOr / visitXor.

**Comment by emgullufsen** [comment]
I see - thank you for the insight I appreciate it (and the --bidirectional tip). I don't want to step on @jayfoad toes, in case he wants to tackle these as well after [D130080](https://reviews.llvm.org/D130080) (and it looks like [D130075](https://reviews.llvm.org/D130075)), but I will watch for those commits and be ready to implement these as and when needed. Thanks again!

**Comment by jayfoad** [comment]
I have no particular plans to work on "or + add" or "xor + add".

+@piotrAMD for awareness.

**Comment by emgullufsen** [comment]
@rotateright since the "outer" operation, or the second function in the composition is the math (add), shouldn't the matching code go in InstCombineAddSub.cpp? In a similar spot to the cannonicalization that [D130080](https://reviews.llvm.org/D130080) is removing (inside foldAddWithConstant method)?
```
(A | C1) + C2 => (A + C2) | C1 iff or doesn't affect same bits as add
(A ^ C1) + C2 => (A + C2) ^ C1 iff xor doesn't affect same bits as add
```
Just a bit confused there - thank you!

**Comment by jayfoad** [comment]
The whole point is to make sure that add-with-constant is the outer operation, in case it is being used as part of a load/store address expression.

**Comment by emgullufsen** [comment]
I see - I was mistaken. So we'll be looking to match on something like:
```
(A + C2) | C
```
and convert to do the bitwise logic first:
```
(A | C) + C2
```
when C and C2 fit conditions. Now the implementation suggestion from @rotateright makes sense, I'd been thinking we were going in reverse direction, sorry for that.

**Comment by emgullufsen** [comment]
Do you guys think something like the below is the general idea for the helper function to be called from visitOr, visitXor, visitAnd?
```
static Instruction *canonicalizeLogicFirst(BinaryOperator &I, 
                                           InstCombiner::BuilderTy &Builder) {
  Type *Ty = I.getType();
  unsigned OpC = I.getOpcode();
  Value *Op0 = I.getOperand(0);
  Value *Op1 = I.getOperand(1);
  Value *X;
  APInt *C, C2;

  if (!(match(Op0, m_OneUse(m_Add(m_Value(X), m_APInt(C2)))) &&
        match(Op1, m_APInt(C))))
    return nullptr;
  
  switch (OpC) {
  case Instruction::And:
    if (!(C->isNegatedPowerOf2() && *C2 == (*C2 & *C)))
      return nullptr;
    break;
  case Instruction::Or:
    if (!(TODO_INSERT_OR_CONDITIONS_ON_CONSTANTS))
      return nullptr
    break;
  case Instruction::Xor:
    if (!(TODO_INSERT_XOR_CONDITIONS_ON_CONSTANTS))
      return nullptr
    break;
  default:
    return nullptr;
  }

  Value *NewBinOp = BinaryOperator::Create(OpC, X, ConstantInt::get(Ty, *C));
  return Builder.CreateAdd(NewBinOp, ConstantInt::get(Ty, *C2))
}
```
Likely the two bottom switch cases would coalesce, since it seems like the constants need to obey the same conditions there, and I haven't quite worked out how to express those constraints, but just wanted to see if I was on track here, and would wait to submit until [D130080](https://reviews.llvm.org/D130080) is merged.

**Comment by rotateright** [comment]
That seems on the right path. You'd use Alive2 to show correctness for all cases.
"*C2 == (*C2 & *C)" could be written as C2->isSubsetOf(*C)? 

Make sure there are regression tests that exercise the limit of whatever constraints are allowed (ie, have a test with the biggest difference in those constants that is allowed and have another test that is one more than that as a negative test).

"NewBinOp" should be created with the Builder, otherwise it might be inserted in the wrong place. We only create the **final** instruction in a pattern directly in InstCombine (because the pass itself will handle the insert, renaming, and updating of worklists - it's awkward, but I think it has always been like that).

**Comment by emgullufsen** [comment]
Cool sounds good - thanks for the heads up I appreciate it. Oh I see I'd missed that regarding creating the final instruction, will include suggestions in patch, with regression tests.

**Comment by jayfoad** [comment]
> Likely the two bottom switch cases would coalesce, since it seems like the constants need to obey the same conditions there, and I haven't quite worked out how to express those constraints,

I think the best simplest constraint (for all three cases) is that all bits affected by the logical op should be lower than the lowest bit affected by the add. This is less constrained than the current And/Add handling because it allows cases that aren't negative powers of two like:
(x + 16) & -5
As well as new cases like:
(x + 16) | 6
(x + 16) ^ 6

**Comment by emgullufsen** [comment]
Ah I see - that makes sense to me. If I understand correctly then do you think something like the below is on track:
```
static Instruction *canonicalizeLogicFirst(BinaryOperator &I, 
                                           InstCombiner::BuilderTy &Builder) {
  Type *Ty = I.getType();
  unsigned OpC = I.getOpcode();
  Value *Op0 = I.getOperand(0);
  Value *Op1 = I.getOperand(1);
  Value *X;
  APInt *C, C2;
  Value *NewBinOp;

  if (!(match(Op0, m_OneUse(m_Add(m_Value(X), m_APInt(C2)))) &&
        match(Op1, m_APInt(C))))
    return nullptr;

  unsigned Width = Ty->getScalarSizeInBits(); // C2->getBitWidth() instead?
  unsigned C2Lz = C2->countLeadingZeros();
  unsigned CLz  = C->countLeadingZeros();
  unsigned CLo  = C->countLeadingOnes();
  unsigned LastOneMath = Width - C2Lz;

  switch (OpC) {
  case Instruction::And:
    if (!(CLo > LastOneMath))
      return nullptr;
    break;
  case Instruction::Xor:
    LLVM_FALLTHROUGH;
  case Instruction::Or:
    if (!(CLz > LastOneMath))
      return nullptr
    break;
  default:
    return nullptr;
  }
  
  NewBinOp = Builder.CreateBinOp(OpC, x, C)
  return BinaryOperator::CreateAdd(NewBinOp, ConstantInt::get(Ty, *C2))
}
```

**Comment by jayfoad** [comment]
Yes that looks pretty good. If you open a phabricator review then we can nit pick over the details!

**Comment by emgullufsen** [comment]
I submitted two patches - one for baseline regression tests and one for changes to InstCombine:
[1 - Baseline tests addition](https://reviews.llvm.org/D131140)
[2 - Changes to InstCombine and regression tests](https://reviews.llvm.org/D131142)

**Comment by emgullufsen** [comment]
addressed by commit [eb1e2b39978c](https://reviews.llvm.org/rGeb1e2b39978c4ea901198d78d57ff70bf9a85b60)
